src/node_types/sequencer.rs

```diff
@@ -93,44 +90,70 @@ impl Sequencer {
             key,
             tree: Arc::new(Mutex::new(IndexedMerkleTree::new_with_size(1024).unwrap())),
-            pending_entries: Arc::new(Mutex::new(Vec::new())),
+            pending_operations: Arc::new(Mutex::new(Vec::new())),
             epoch_buffer_tx: Arc::new(tx),
             epoch_buffer_rx: Arc::new(Mutex::new(rx)),
         })
     }
 
-    // main_loop is responsible for finalizing epochs every epoch length and writing them to the buffer for DA submission.
-    async fn main_loop(self: Arc<Self>) -> Result<(), tokio::task::JoinError> {
-        info!("starting main sequencer loop");
+    // sync_loop is responsible for downloading operations from the DA layer
+    async fn sync_loop(self: Arc<Self>) -> Result<(), tokio::task::JoinError> {
+        info!("starting operation sync loop");
+        // todo: make self.start_height
+        let start_height = 1000;
         let epoch_buffer = self.epoch_buffer_tx.clone();
-        let mut ticker = interval(Duration::from_secs(self.epoch_duration));
         spawn(async move {
+            let mut current_position = start_height;
             loop {
-                ticker.tick().await;
-                match self.finalize_epoch().await {
-                    Ok(epoch) => {
-                        let epoch_height = match self.db.get_epoch() {
-                            Ok(epoch) => epoch,
-                            Err(e) => {
-                                error!("sequencer_loop: getting epoch from db: {}", e);
-                                continue;
+                // target is updated when a new header is received
+                let target = match self.da.get_latest_height().await {
+                    Ok(target) => target,
+                    Err(e) => {
+                        error!("failed to update sync target, retrying: {:?}", e);
+                        continue;
+                    }
+                };
+
+                debug!("updated sync target to height {}", target);
+                while current_position < target {
+                    trace!("processing height: {}", current_position);
+                    match self.da.get_operations(current_position + 1).await {
+                        Ok(operations) => {
+                            if !operations.is_empty() {
+                                debug!(
+                                    "sequencer: got operations at height {}",
+                                    current_position + 1
+                                );
                             }
-                        };
 
-                        info!("sequencer_loop: finalized epoch {}", epoch_height);
-                        match epoch_buffer.send(epoch).await {
-                            Ok(_) => (),
-                            Err(e) => {
-                                error!("sequencer_loop: sending epoch to buffer: {}", e);
+                            // TODO: doesn't this copy the Vec? Seems very inefficient
```

Comment:
Hmh, no i dont think its copying the Vec. the finalize_epoch method takes ownership of the Vec and afaik its moved, not copied and if i remember correctly its not an expensive operation in Rust. But i also think we can use &Vec<Operation> to make clear, that the functions doesnt even need to take ownership of the vec.

Reply:
kk ill remove the todo